File: hash-table/subarray-sums-divisible-by-k/subarray-sums-divisible-by-k.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
function subarraysDivByK (nums, k) {
  const freq = new Array(k).fill(0);
  freq[0] = 1;
 
  let sum = 0;
  return nums.reduce((acc, num) => {
    sum += num;

    let remainder = sum % k;

    // only consider positive remainders
    if (remainder < 0) {
      remainder += k;
    }
    
    acc += freq[remainder];

    freq[remainder]++;
        
    return acc;
  }, 0);
}

File: hash-table/subarray-sums-divisible-by-k/subarray-sums-divisible-by-k.go
func subarraysDivByK(nums []int, k int) int {
	freq := make([]int, k)
	sum := 0
	res := 0
	freq[0] = 1

	for _, num := range nums {
		sum += num
		remainder := sum % k

		if remainder < 0 {
			remainder += k
		}

		res += freq[remainder]
		freq[remainder]++
	}

	return res
}
